File: src/types/resume.ts
/**
 * Strongly-typed model for the résumé data that powers every command.
 * All values are sourced directly from the attached PDF — nothing is invented.
 * Optional fields are omitted when the source résumé does not provide them.
 */

export interface SocialLink {
  /** Handle as shown on the résumé, e.g. "sivaneshb". */
  handle: string;
  /** Fully-qualified URL. */
  url: string;
}

export interface ContactInfo {
  email: string;
  phone?: string;
  location: string;
  linkedin?: SocialLink;
  github?: SocialLink;
  portfolio?: SocialLink;
}

export interface SkillCategory {
  /** Category label as grouped on the résumé. */
  name: string;
  /** Short slug used for the ASCII underline + search. */
  skills: string[];
}

export interface ExperienceItem {
  role: string;
  company: string;
  /** Human-readable start, e.g. "08/2024". */
  start: string;
  /** Human-readable end, e.g. "Present". */
  end: string;
  location: string;
  highlights: string[];
}

export interface Certification {
  name: string;
  issuer?: string;
}

export interface EducationItem {
  degree: string;
  institution: string;
  start: string;
  end: string;
  location?: string;
}

export interface Project {
  name: string;
  description: string;
  tech: string[];
}

export interface TimelineEvent {
  /** Sort key + label, e.g. "2024" or "Aug 2024". */
  date: string;
  title: string;
  subtitle?: string;
}

export interface ResumeData {
  name: string;
  title: string;
  /** Shell username shown in the prompt: `<username>@<host>:~$`. */
  username: string;
  host: string;
  summary: string;
  contact: ContactInfo;
  skills: SkillCategory[];
  experience: ExperienceItem[];
  certifications: Certification[];
  education: EducationItem[];
  /** Derived strictly from experience bullets — real initiatives, not invented. */
  projects: Project[];
  /** Quantified wins pulled from experience bullets. */
  achievements: string[];
  timeline: TimelineEvent[];
  /** ISO date the current role started — used to compute years of experience. */
  careerStartISO: string;
  /** Path (relative to the site base) to the downloadable PDF. */
  resumeFile: string;
}

File: ssh/internal/resume/resume.go
// Package resume defines the portfolio data model and loads it at runtime from
// the shared, single-source-of-truth file (shared/resume.json). The same file
// is consumed by the React website, so editing it updates both frontends.
package resume

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

type SocialLink struct {
	Handle string `json:"handle"`
	URL    string `json:"url"`
}

type Contact struct {
	Email     string      `json:"email"`
	Phone     string      `json:"phone,omitempty"`
	Location  string      `json:"location"`
	LinkedIn  *SocialLink `json:"linkedin,omitempty"`
	GitHub    *SocialLink `json:"github,omitempty"`
	Portfolio *SocialLink `json:"portfolio,omitempty"`
}

type SkillCategory struct {
	Name   string   `json:"name"`
	Skills []string `json:"skills"`
}

type Experience struct {
	Role       string   `json:"role"`
	Company    string   `json:"company"`
	Start      string   `json:"start"`
	End        string   `json:"end"`
	Location   string   `json:"location"`
	Highlights []string `json:"highlights"`
}

type Certification struct {
	Name   string `json:"name"`
	Issuer string `json:"issuer,omitempty"`
}

type Education struct {
	Degree      string `json:"degree"`
	Institution string `json:"institution"`
	Location    string `json:"location,omitempty"`
	Start       string `json:"start"`
	End         string `json:"end"`
}

type Project struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Tech        []string `json:"tech"`
}

type TimelineEvent struct {
	Date     string `json:"date"`
	Title    string `json:"title"`
	Subtitle string `json:"subtitle,omitempty"`
}

// Resume mirrors shared/resume.json field-for-field.
type Resume struct {
	Name           string          `json:"name"`
	Title          string          `json:"title"`
	Username       string          `json:"username"`
	Host           string          `json:"host"`
	Summary        string          `json:"summary"`
	Contact        Contact         `json:"contact"`
	Skills         []SkillCategory `json:"skills"`
	Experience     []Experience    `json:"experience"`
	Certifications []Certification `json:"certifications"`
	Education      []Education     `json:"education"`
	Projects       []Project       `json:"projects"`
	Achievements   []string        `json:"achievements"`
	Timeline       []TimelineEvent `json:"timeline"`
	CareerStartISO string          `json:"careerStartISO"`
	ResumeFile     string          `json:"resumeFile"`
}

// candidatePaths lists where the shared résumé JSON may live, in priority order.
func candidatePaths(explicit string) []string {
	paths := []string{}
	if explicit != "" {
		paths = append(paths, explicit)
	}
	if env := os.Getenv("RESUME_PATH"); env != "" {
		paths = append(paths, env)
	}
	return append(paths,
		"shared/resume.json",
		"../shared/resume.json",
		"../../shared/resume.json",
		"../../../shared/resume.json",
		"/app/shared/resume.json",
	)
}

// Load reads and parses the résumé from the first path that exists.
func Load(explicit string) (*Resume, error) {
	tried := candidatePaths(explicit)
	for _, p := range tried {
		if p == "" {
			continue
		}
		b, err := os.ReadFile(p)
		if err != nil {
			continue
		}
		return Parse(b)
	}
	return nil, fmt.Errorf("resume.json not found (looked in %v); set RESUME_PATH", tried)
}

// Parse decodes résumé JSON and validates the essentials. Unknown fields are
// tolerated so the shared file can grow web-only sections without breaking SSH.
func Parse(b []byte) (*Resume, error) {
	var r Resume
	if err := json.Unmarshal(b, &r); err != nil {
		return nil, fmt.Errorf("parse resume.json: %w", err)
	}
	if r.Name == "" || len(r.Skills) == 0 || len(r.Experience) == 0 {
		return nil, fmt.Errorf("resume.json is missing required fields (name/skills/experience)")
	}
	return &r, nil
}

// AllTechnologies returns the de-duplicated union of every skill.
func (r *Resume) AllTechnologies() []string {
	seen := map[string]bool{}
	var out []string
	for _, c := range r.Skills {
		for _, s := range c.Skills {
			if !seen[s] {
				seen[s] = true
				out = append(out, s)
			}
		}
	}
	return out
}

// ExperienceMonths returns whole months of experience since CareerStartISO.
func (r *Resume) ExperienceMonths(now time.Time) int {
	start, err := time.Parse("2006-01-02", r.CareerStartISO)
	if err != nil {
		return 0
	}
	months := (now.Year()-start.Year())*12 + int(now.Month()) - int(start.Month())
	if now.Day() < start.Day() {
		months--
	}
	if months < 0 {
		months = 0
	}
	return months
}

// ExperienceLabel renders a human duration like "2 yrs 1 mo".
func (r *Resume) ExperienceLabel(now time.Time) string {
	m := r.ExperienceMonths(now)
	years, months := m/12, m%12
	label := ""
	if years > 0 {
		label += fmt.Sprintf("%d yr", years)
		if years > 1 {
			label += "s"
		}
		label += " "
	}
	label += fmt.Sprintf("%d mo", months)
	if months != 1 {
		label += "s"
	}
	return label
}

// Employers counts distinct companies in the experience list.
func (r *Resume) Employers() int {
	seen := map[string]bool{}
	for _, e := range r.Experience {
		seen[e.Company] = true
	}
	return len(seen)
}
